### PR TITLE
Issue #SH-12: Enhancements in Score report to add new fields

### PR DESCRIPTION
**Pull request by utk14**

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 1]
```diff
@@ -133,30 +151,34 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
         col("maskedphone"),
         col("rootorgid"),
         col("locationids"),
-        col("channel")
-      ).cache()
+        col("channel"),
+        col("locationids")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:155]: col('locationids') is duplicated?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 2]
```diff
@@ -133,30 +151,34 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
+      ).persist()
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:156]: for persist - provide MEM_AND_DISK as explicit option

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 3]
```diff
@@ -133,30 +151,34 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
-        && externalIdentityDF.col("provider") === userDF.col("channel")
-        && externalIdentityDF.col("userid") === userDF.col("userid"), "inner")
-      .select(externalIdentityDF.col("externalid"), externalIdentityDF.col("userid"))
+      .select(col("provider"), col("idtype"), col("externalid"), col("userid")).persist()
+
+    val systemSettingDF = loadData(spark, Map("table" -> "system_settings", "keyspace" -> sunbirdKeyspace))
+      .where(col("id") === "custodianOrgId" && col("field") === "custodianOrgId")
+      .select(col("value")).persist()
+
+    val custRootOrgId = systemSettingDF.select("value").first().getString(0)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:177]: Move this to a method named - getCustodianOrgId()

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 4]
```diff
@@ -133,30 +151,34 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
+    val detailsByUser = getUserSelfDeclaredDetails(userDF, custRootOrgId, externalIdentityDF)
+    val detailsByState = getStateDeclaredDetails(userDF, custRootOrgId, externalIdentityDF, organisationDF, userOrgDF)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:179]: Please unpersist the dataframe once the usage is over.

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 5]
```diff
@@ -191,65 +213,60 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
+
+    val stateInfoByOrg = locationInfoByOrg.join(userDF,
+      locationInfoByOrg.col("id") === userDF.col("rootorgid") &&
+        locationInfoByOrg.col("isrootorg").equalTo(true))
+      .select(col("state_name"),
+        userDF.col("userid"))
+
+    val resolvedStateDenormDF = stateInfoByUser.union(stateInfoByOrg)
+        .dropDuplicates(Seq("userid"))
+
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:245]: What is the purpose of this logic?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 6]
```diff
@@ -191,65 +213,60 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
-      .union(schoolNameIndexDF.filter(col("index") =!= 1).groupBy("userid").agg(collect_list("schoolname_resolved").cast("string").as("schoolname_resolved")))
-
-    resolvedExternalIdDF
-      .join(resolvedSchoolNameDF, Seq("userid"), "left_outer")
+    val finalReportDF = userStateResolvedDF
+      .join(userInfoDF, Seq("userid"), "left_outer")
       .join(resolvedOrgNameDF, Seq("userid", "rootorgid"), "left_outer")
       .dropDuplicates("userid")
-      .cache();
+      finalReportDF.cache();
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:261]: Do a persist(MEM_AND_DISK_ONLY) option

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 7]
```diff
@@ -191,65 +213,60 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
-
-    val resolvedSchoolNameDF = schoolNameIndexDF.selectExpr("*").filter(col("index") === 1).drop("organisationid", "index", "batchid")
-      .union(schoolNameIndexDF.filter(col("index") =!= 1).groupBy("userid").agg(collect_list("schoolname_resolved").cast("string").as("schoolname_resolved")))
-
-    resolvedExternalIdDF
-      .join(resolvedSchoolNameDF, Seq("userid"), "left_outer")
+    val finalReportDF = userStateResolvedDF
+      .join(userInfoDF, Seq("userid"), "left_outer")
       .join(resolvedOrgNameDF, Seq("userid", "rootorgid"), "left_outer")
       .dropDuplicates("userid")
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:260]: Do we still have duplicates at this point?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/AssessmentMetricsJob.scala [hunk 8]
```diff
@@ -130,19 +147,40 @@ object AssessmentMetricsJob extends optional.Application with IJob with BaseRepo
...
+
+    val externalIDByUserDF = externalIdentityDF.join(organisationDF,
+      externalIdentityDF.col("provider") === organisationDF.col("channel") &&
+        externalIdentityDF.col("idtype").equalTo("declared-ext-id"), "inner")
+      .join(userDenormDF, Seq("userid"), "inner")
+      .withColumn("externalid_resolved", when(userDenormDF.col("course_channel") === userDenormDF.col("channel"), externalIdentityDF.col("externalid")).otherwise(""))
+      .select(col("externalid_resolved"), externalIdentityDF.col("userid"))
+
+    val externalIdMapDF = externalIdByStateDF.union(externalIDByUserDF)
+      .dropDuplicates(Seq("userid"))
```
[the remark this answers is not in the document]
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/AssessmentMetricsJob.scala]: I have made the changes for the AssessmentMetricsJob as well

File: data-products/src/main/scala/org/sunbird/analytics/job/report/AssessmentMetricsJob.scala [hunk 9]
```diff
@@ -2,18 +2,21 @@ package org.sunbird.analytics.job.report
...
+import org.ekstep.analytics.framework._
+import org.ekstep.analytics.framework.fetcher.DruidDataFetcher
 import org.ekstep.analytics.framework.util.DatasetUtil.extensions
 import org.ekstep.analytics.framework.util.{CommonUtil, JSONUtils, JobLogger}
 import org.joda.time.DateTime
 import org.joda.time.format.DateTimeFormat
 import org.sunbird.analytics.util.ESUtil
 import org.sunbird.cloud.storage.conf.AppConf
 
+case class druidOutput(identifier: String, channel: String)
```
[the remark this answers is not in the document]
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/AssessmentMetricsJob.scala:18]: Done

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala (hunk 7 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:260]: Duplicate records are coming if we remove the drop duplicate from this part of the code

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala [hunk 10]
```diff
@@ -134,29 +153,32 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
...
-      .join(externalIdentityDF, externalIdentityDF.col("idtype") === userDF.col("channel")
-        && externalIdentityDF.col("provider") === userDF.col("channel")
-        && externalIdentityDF.col("userid") === userDF.col("userid"), "inner")
-      .select(externalIdentityDF.col("externalid"), externalIdentityDF.col("userid"))
+      .select(col("provider"), col("idtype"), col("externalid"), col("userid")).persist()
+
+    val systemSettingDF = loadData(spark, Map("table" -> "system_settings", "keyspace" -> sunbirdKeyspace))
+      .where(col("id") === "custodianOrgId" && col("field") === "custodianOrgId")
+      .select(col("value")).persist()
+
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:176]: move this code to getCustodianOrgId() method as well

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala (hunk 5 above)
Comment (SanthoshVasabhaktula) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:245]: Can we move this to same method - getUserSelfDeclaredDetails() & getStateDeclaredDetails()?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala (hunk 7 above)
Comment (SanthoshVasabhaktula) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:260]: Which code is causing the duplicates?

File: data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala (hunk 5 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala:245]: Done

**Pull request merged**